File: articles.jdbc/src/main/java/pet/db/jdbc/service/RegistrationServiceImpl.java
package pet.db.jdbc.service;

import lombok.RequiredArgsConstructor;

import org.springframework.security.crypto.password.PasswordEncoder;

import org.springframework.stereotype.Service;

import pet.db.jdbc.model.dto.User;
import pet.db.jdbc.tool.exception.DuplicateUserException;

@Service
@RequiredArgsConstructor
public class RegistrationServiceImpl implements RegistrationService {

    private final UserService userService;

    private final PasswordEncoder passwordEncoder;

    @Override
    public User register(User user) throws DuplicateUserException {
        User UserForRegistration = buildUserForRegistration(user);
        return userService.create(UserForRegistration);
    }

    private User buildUserForRegistration(User user) {
        String encodedPassword = passwordEncoder.encode(user.getPassword());
        return user.toBuilder()
                .password(encodedPassword)
                .build();
    }

}

File: src/main/kotlin/pet/articles/service/RegistrationServiceImpl.kt
package pet.articles.service

import org.koin.core.annotation.Single

import org.mindrot.jbcrypt.BCrypt

import pet.articles.model.dto.User

class RegistrationServiceImpl(
    private val userService: UserService,
    private val bcryptSalt: String
) : RegistrationService {

    override fun register(user: User): User {
        val userForRegistration: User = setEncryptedPassword(user)
        return userService.create(userForRegistration)
    }

    private fun setEncryptedPassword(user: User): User =
        user.copy(password = BCrypt.hashpw(user.password, bcryptSalt))
}